### PR TITLE
feat: create catalog by sql

### PR DESCRIPTION
**Pull request by ClSlaid**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Try to implement `CREATE CATALOG` SQL in databend-query

### CODE REVIEW

File: src/meta/app/src/schema/catalog.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        self.meta.catalog_type
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct CatalogMeta {
+    pub catalog_type: CatalogType,
+    pub options: BTreeMap<String, String>,
+    pub created_on: Option<DateTime<Utc>>,
+    pub droped_on: Option<DateTime<Utc>>,
```
Comment (Xuanwo) [inline src/meta/app/src/schema/catalog.rs:87]: typo here, should be `dropped_on`.

**Comment by Xuanwo** [comment]
Any updates on this?

**Comment by ClSlaid** [comment]
> Any updates on this?

Kind of busy recently, still working on it.

File: src/meta/api/src/schema_api_impl.rs [hunk 2]
```diff
@@ -145,6 +153,112 @@ const DEFAULT_DATA_RETENTION_SECONDS: i64 = 24 * 60 * 60;
...
+                        ),
+                    )))
+                };
+            }
+
+            // get ctl id list from _fd_ctl_id_list/ctl_id
+            let ctl_id_list_key = CatalogIdListKey {
+                tenant: name_key.tenant.clone(),
+                ctl_name: name_key.ctl_name.clone(),
+            };
```
Comment (drmingdrmer) [inline src/meta/api/src/schema_api_impl.rs:193]: Are you going to support time travel for the catalog API?

File: src/meta/app/src/schema/catalog.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            CatalogType::Default => write!(f, "DEFAULT"),
+            CatalogType::Hive => write!(f, "HIVE"),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct CatalogNameIdent {
+    pub tenant: String,
+    pub ctl_name: String,
```
Comment (drmingdrmer) [inline src/meta/app/src/schema/catalog.rs:61]: I'd recommend using a whole word for a field name: `ctl_name`: `catalog_name`

File: src/meta/app/src/schema/catalog.rs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+pub struct CatalogInfo {
+    pub name_ident: CatalogNameIdent,
+    pub meta: CatalogMeta,
+}
+
+impl CatalogInfo {
+    pub fn catalog_type(&self) -> CatalogType {
+        self.meta.catalog_type
+    }
+}
```
Comment (drmingdrmer) [inline src/meta/app/src/schema/catalog.rs:80]: Remove this if this is just an analog of `TableInfo`. `TableInfo` is kept for only historical reasons.

File: src/meta/app/src/schema/catalog.rs [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct CatalogMeta {
+    pub catalog_type: CatalogType,
+    pub options: BTreeMap<String, String>,
+    pub created_on: DateTime<Utc>,
+    pub dropped_on: Option<DateTime<Utc>>,
```
Comment (drmingdrmer) [inline src/meta/app/src/schema/catalog.rs:87]: Does a catalog need `dropped_on` time?

Comment (ClSlaid) [reply src/meta/app/src/schema/catalog.rs:87]: The user-defined catalog will support dropping in the future.

Comment (drmingdrmer) [reply src/meta/app/src/schema/catalog.rs:87]: Is it ever possible for a user to see this time?

File: src/meta/api/src/schema_api_impl.rs (hunk 2 above)
Comment (ClSlaid) [reply src/meta/api/src/schema_api_impl.rs:193]: Nope

**Comment by Xuanwo** [comment]
This PR is too large that needs the whole databend team to review it. I prefer to split it into serval parts so that we can add this feature gradually.

**Pull request closed without merging**